pints/toy/_stochastic_degradation_model.py

```diff
@@ -0,0 +1,90 @@
+#
+# Logistic toy model.
+#
+# This file is part of PINTS.
+#  Copyright (c) 2017-2019, University of Oxford.
+#  For licensing information, see the LICENSE file distributed with the PINTS
+#  software package.
+#
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+import numpy as np
+import scipy
+from scipy.interpolate import interp1d
+import pints
+
+from . import ToyModel
+
+
+class StochasticDegradationModel(pints.ForwardModel, ToyModel):
+
+    """
+    Stochastic decay model of a single chemical reaction [1].
+
+    Time until next reaction...
+    .. math::
+    $\tau = \frac{1}{A(t)k}*\ln{\frac{1}{r}}$
+
+    Has one parameter: Rate constant :math:`k`.
+    :math:`r` is a random variable, which is part of the stochastic model
+    The initial concentration :math:`A(0) = n_0` can be set using the
+    (optional) named constructor arg ``initial_concentration``
+
+    [1] Erban et al., 2007
+
+    *Extends:* :class:`pints.ForwardModel`, :class:`pints.toy.ToyModel`.
+    """
+
+    def __init__(self, initial_concentration=20):
+        super(StochasticDegradationModel, self).__init__()
+        self._n0 = float(initial_concentration)
+        if self._n0 < 0:
+            raise ValueError('Initial concentration cannot be negative.')
+
+    def n_parameters(self):
+        """ See :meth:`pints.ForwardModel.n_parameters()`. """
+        return 1
+
+    def simulate(self, parameters, times):
+        """ See :meth:`pints.ForwardModel.simulate()`. """
+        if parameters <= 0:
+            raise ValueError('rate constant must be positive')
+
+        k = np.array([float(parameters)])
+        times = np.asarray(times)
+        if np.any(times < 0):
+            raise ValueError('Negative times are not allowed.')
+        if self._n0 == 0:
+            return np.zeros(times.shape)
+        a = np.array([float(self._n0)])
+
+        t = 0
+        mol_conc = [a]
+        time = [t]
+
+        # Run stochastic degradation algorithm, calculating time until next
+        # reaction and decreasing concentration by 1 at that time
+        while a > 0:
+            r = np.random.uniform(0, 1)
+            tao = ((1 / (a * k)) * np.log(1 / r))[0]
+            t += tao
+            time.append(t)
+            a = a - 1
+            mol_conc.append(a[0])
+
+        # Interpolate as step function, decreasing mol_conc by 1 at each
+        # reaction time point
+        f1 = interp1d(time, mol_conc, kind='previous')
+
+        # Compute concentration ('a') values at given time points using f1
+        # at any time beyond the last reaction, concentration = 0
+        values = f1(times[np.where(times <= max(time))])
+        zero_vector = np.zeros(len(times[np.where(times > max(time))]))
+        values = np.concatenate((values, zero_vector))
+
```

Comment:
As Chon suggested when we met, I really think it'd be a good idea to have extra functions for this model to return useful things. For example, a function which returns the deterministic mean and variance at a given point in time.

I am less sure how to handle the exact posterior here; whether it's best to have a separate toy distribution (rather than toy model) which returns the likelihood of the data or whether to include it here? For clarity, what this function would do is take a series of data - times and values - and return their probability given the rate parameter k. To convert this into a Bayesian problem, we would need to combine this with a prior on k.